File: src/test/java/com/digitalsanctuary/spring/demo/user/ui/page/DeleteAccountPage.java
package com.digitalsanctuary.spring.demo.user.ui.page;

import com.codeborne.selenide.Condition;
import com.codeborne.selenide.Selenide;
import com.codeborne.selenide.SelenideElement;

import static com.codeborne.selenide.Condition.*;

/**
 * Page object for delete account page
 */
public class DeleteAccountPage {

    private final SelenideElement DELETE_ACCOUNT_BUTTON = Selenide.$x("//button[@type='submit']");
    private final SelenideElement DELETE_CONFIRMATION_MODAL = Selenide.$x("//div[@id='deleteConfirmationModal']");
    private final SelenideElement CONFIRMATION_INPUT = Selenide.$x("//input[@id='deleteConfirmationInput']");
    private final SelenideElement CONFIRM_DELETION_BUTTON = Selenide.$x("//button[@id='confirmDeletionButton']");
    private final SelenideElement CANCEL_BUTTON = Selenide.$x("//button[contains(@class, 'btn-secondary')]");
    private final SelenideElement GLOBAL_MESSAGE = Selenide.$x("//div[@id='globalMessage']");
    private final SelenideElement GLOBAL_ERROR = Selenide.$x("//div[@id='globalError']");

    public DeleteAccountPage(String url) {
        Selenide.open(url);
    }

    public DeleteAccountPage() {
        // For navigation to this page from other pages
    }

    /**
     * Start the account deletion process by clicking the delete button
     */
    public DeleteAccountPage startDeletion() {
        DELETE_ACCOUNT_BUTTON.click();
        // Wait for modal to appear
        DELETE_CONFIRMATION_MODAL.should(appear);
        return this;
    }

    /**
     * Complete account deletion with confirmation
     */
    public DeleteAccountPage confirmDeletion(String confirmationText) {
        CONFIRMATION_INPUT.setValue(confirmationText);
        CONFIRM_DELETION_BUTTON.click();
        return this;
    }

    /**
     * Complete account deletion with correct confirmation text
     */
    public DeleteAccountPage confirmDeletion() {
        return confirmDeletion("DELETE");
    }

    /**
     * Cancel the deletion process
     */
    public DeleteAccountPage cancelDeletion() {
        CANCEL_BUTTON.click();
        // Wait for modal to disappear
        DELETE_CONFIRMATION_MODAL.should(disappear);
        return this;
    }

    /**
     * Get success message after deletion
     */
    public String getSuccessMessage() {
        return GLOBAL_MESSAGE.shouldBe(visible).text();
    }

    /**
     * Get error message
     */
    public String getErrorMessage() {
        return GLOBAL_ERROR.shouldBe(visible).text();
    }

    /**
     * Check if deletion was successful
     */
    public boolean isDeletionSuccessful() {
        return GLOBAL_MESSAGE.has(cssClass("alert-success"));
    }

    /**
     * Check if there's an error message
     */
    public boolean hasError() {
        return GLOBAL_ERROR.exists() && GLOBAL_ERROR.isDisplayed();
    }

    /**
     * Check if the confirmation modal is visible
     */
    public boolean isConfirmationModalVisible() {
        return DELETE_CONFIRMATION_MODAL.has(cssClass("show"));
    }

    /**
     * Check if the delete form is hidden (after successful deletion)
     */
    public boolean isDeleteFormHidden() {
        SelenideElement form = Selenide.$x("//form[@id='deleteAccountForm']");
        return form.has(cssClass("d-none"));
    }

    /**
     * Wait for page to load completely
     */
    public DeleteAccountPage waitForPageLoad() {
        DELETE_ACCOUNT_BUTTON.should(appear);
        return this;
    }

    /**
     * Clear confirmation input
     */
    public DeleteAccountPage clearConfirmationInput() {
        CONFIRMATION_INPUT.clear();
        return this;
    }

    /**
     * Get the current confirmation input value
     */
    public String getConfirmationInputValue() {
        return CONFIRMATION_INPUT.getValue();
    }
}

File: playwright/src/pages/DeleteAccountPage.ts
import { Page, Locator, expect } from '@playwright/test';
import { BasePage } from './BasePage';

/**
 * Page Object for the Delete Account page.
 */
export class DeleteAccountPage extends BasePage {
  readonly path = '/user/delete-account.html';

  // Form elements
  readonly deleteButton: Locator;

  // Modal elements
  readonly confirmationModal: Locator;
  readonly confirmationInput: Locator;
  readonly confirmButton: Locator;
  readonly cancelButton: Locator;

  // Messages
  readonly globalMessage: Locator;
  readonly globalError: Locator;

  constructor(page: Page) {
    super(page);
    this.deleteButton = page.locator('#deleteAccountForm button[type="submit"]');
    this.confirmationModal = page.locator('#deleteConfirmationModal');
    this.confirmationInput = page.locator('#deleteConfirmationInput');
    this.confirmButton = page.locator('#confirmDeletionButton');
    this.cancelButton = page.locator('.modal-footer button[data-bs-dismiss="modal"]');
    this.globalMessage = page.locator('#globalMessage');
    this.globalError = page.locator('#globalError');
  }

  /**
   * Click delete account button to show confirmation modal.
   */
  async clickDelete(): Promise<void> {
    await this.deleteButton.click();
  }

  /**
   * Wait for modal to be visible.
   */
  async waitForModal(): Promise<void> {
    await this.confirmationModal.waitFor({ state: 'visible' });
  }

  /**
   * Type confirmation text in modal.
   */
  async typeConfirmation(text: string): Promise<void> {
    await this.confirmationInput.fill(text);
  }

  /**
   * Confirm deletion in modal.
   */
  async confirmDeletion(): Promise<void> {
    await this.confirmButton.click();
  }

  /**
   * Cancel deletion and close modal.
   */
  async cancelDeletion(): Promise<void> {
    await this.cancelButton.click();
  }

  /**
   * Complete the account deletion flow.
   */
  async deleteAccount(): Promise<void> {
    await this.clickDelete();
    await this.waitForModal();
    await this.typeConfirmation('DELETE');
    await this.confirmDeletion();
  }

  /**
   * Delete account and wait for success message.
   * The app uses AJAX and shows a success message instead of redirecting.
   * The session is invalidated server-side.
   */
  async deleteAccountAndWait(): Promise<void> {
    await this.deleteAccount();
    // Wait for the success message to appear (AJAX response)
    await this.globalMessage.waitFor({ state: 'visible', timeout: 10000 });
    // Also wait for the form to be hidden (confirmation of success)
    await this.page.waitForFunction(() => {
      const form = document.querySelector('#deleteAccountForm');
      return form?.classList.contains('d-none');
    }, { timeout: 5000 });
  }

  /**
   * Check if modal is visible.
   */
  async isModalVisible(): Promise<boolean> {
    return this.confirmationModal.isVisible();
  }

  /**
   * Check if error message is displayed.
   */
  async hasError(): Promise<boolean> {
    return this.globalError.isVisible();
  }

  /**
   * Get error message text.
   */
  async getErrorText(): Promise<string | null> {
    if (await this.hasError()) {
      return this.globalError.textContent();
    }
    return null;
  }

  /**
   * Check if success message is displayed.
   */
  async hasSuccessMessage(): Promise<boolean> {
    return this.globalMessage.isVisible();
  }
}
